File: src/cstlib/synaptic.py
"""Canonical CST Synaptic Function v1.

This module is intentionally dependency-free and mirrors the reference C++, Rust,
JavaScript, Go, Java, C#, Swift, Kotlin and C ABI implementations.
"""
from __future__ import annotations

from dataclasses import dataclass
from math import exp, tanh
from typing import Sequence


def gaussian_affinity(a: Sequence[float], b: Sequence[float], sigma: float = 1.0) -> float:
    """Return exp(-||a-b||^2 / (2*sigma^2))."""
    if sigma <= 0:
        raise ValueError("sigma must be > 0")
    if len(a) != len(b) or not a:
        raise ValueError("a and b must be non-empty and have equal length")
    distance_sq = sum((float(x) - float(y)) ** 2 for x, y in zip(a, b))
    return exp(-distance_sq / (2.0 * sigma * sigma))


def affinity_matrix(states: Sequence[Sequence[float]], sigma: float = 1.0) -> list[list[float]]:
    if not states:
        raise ValueError("states must be non-empty")
    dimension = len(states[0])
    if dimension == 0 or any(len(row) != dimension for row in states):
        raise ValueError("all states must be non-empty and have equal dimension")
    return [[gaussian_affinity(a, b, sigma) for b in states] for a in states]


def gated_blend(standard: float, state_affinity: float, gate: float) -> float:
    if not 0.0 <= gate <= 1.0:
        raise ValueError("gate must be in [0, 1]")
    return (1.0 - gate) * float(standard) + gate * float(state_affinity)


def state_step(state: Sequence[float], signal: Sequence[float], *, decay: float = 0.92, gain: float = 1.0, dt: float = 1.0) -> list[float]:
    if len(state) != len(signal) or not state:
        raise ValueError("state and signal must be non-empty and have equal length")
    if not 0.0 <= decay <= 1.0:
        raise ValueError("decay must be in [0, 1]")
    if dt < 0:
        raise ValueError("dt must be >= 0")
    effective_decay = decay**dt
    return [effective_decay * float(x) + (1.0 - effective_decay) * float(gain) * tanh(float(u)) for x, u in zip(state, signal)]


@dataclass(frozen=True)
class SynapticFunction:
    sigma: float = 1.0
    gate: float = 0.5

    def __post_init__(self) -> None:
        if self.sigma <= 0:
            raise ValueError("sigma must be > 0")
        if not 0.0 <= self.gate <= 1.0:
            raise ValueError("gate must be in [0, 1]")

    def affinity(self, a: Sequence[float], b: Sequence[float]) -> float:
        return gaussian_affinity(a, b, self.sigma)

    def matrix(self, states: Sequence[Sequence[float]]) -> list[list[float]]:
        return affinity_matrix(states, self.sigma)

    def blend(self, standard: float, state_affinity: float, gate: float | None = None) -> float:
        return gated_blend(standard, state_affinity, self.gate if gate is None else gate)

    def step(self, state: Sequence[float], signal: Sequence[float], *, decay: float = 0.92, gain: float = 1.0, dt: float = 1.0) -> list[float]:
        return state_step(state, signal, decay=decay, gain=gain, dt=dt)

File: bindings/go/synaptic/synaptic.go
package synaptic

import("errors";"math")
type Function struct{Sigma,Gate float64}
func New(sigma,gate float64)(Function,error){if sigma<=0{return Function{},errors.New("sigma must be > 0")};if gate<0||gate>1{return Function{},errors.New("gate must be in [0,1]")};return Function{Sigma:sigma,Gate:gate},nil}
func GaussianAffinity(a,b[]float64,sigma float64)(float64,error){if sigma<=0||len(a)==0||len(a)!=len(b){return 0,errors.New("invalid vectors or sigma")};var d2 float64;for i:=range a{d:=a[i]-b[i];d2+=d*d};return math.Exp(-d2/(2*sigma*sigma)),nil}
func(f Function)Affinity(a,b[]float64)(float64,error){return GaussianAffinity(a,b,f.Sigma)}
func(f Function)Blend(s,a float64,gate *float64)(float64,error){g:=f.Gate;if gate!=nil{g=*gate};if g<0||g>1{return 0,errors.New("gate must be in [0,1]")};return(1-g)*s+g*a,nil}
func StateStep(state,signal[]float64,decay,gain,dt float64)([]float64,error){if len(state)==0||len(state)!=len(signal)||decay<0||decay>1||dt<0{return nil,errors.New("invalid state update arguments")};ed:=math.Pow(decay,dt);out:=make([]float64,len(state));for i:=range state{out[i]=ed*state[i]+(1-ed)*gain*math.Tanh(signal[i])};return out,nil}
func(f Function)Step(state,signal[]float64,decay,gain,dt float64)([]float64,error){return StateStep(state,signal,decay,gain,dt)}
